src/Microsoft.Identity.Client/ConfidentialClientApplication.cs

```diff
@@ -79,7 +88,13 @@ static ConfidentialClientApplication()
         /// enables app developers to specify the authority
         public ConfidentialClientApplication(string clientId, string redirectUri,
             ClientCredential clientCredential, TokenCache userTokenCache, TokenCache appTokenCache)
-            : this(clientId, DefaultAuthority, redirectUri, clientCredential, userTokenCache, appTokenCache)
+            : this(ConfidentialClientApplicationBuilder.CreateWithApplicationOptions(new ApplicationOptions { ClientId = clientId })
```

Comment:
How do you feel about ApplicationOptions ctor taking in the clientId to enforce that it is a mandatory param? I believe this is an ok modification of the builder pattern. Otherwise you have to throw from .BuildConfiguration(), which is also fine, but we inform the user at runtime.

Reply:
ApplicationOptions is meant to be deserialized from config.  I now have a Create(clientId) method on the Builder classes to ensure we get a clientid and then we'll validate at build time that we have enough of the other parameters set to proceed.  I don't have the validation logic written yet because I want to stabilize the API surface and then do validation and write a suite of unit tests to cover the validation cases.

---
In reply to: [241521301](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/723#discussion_r241521301) [](ancestors = 241521301)